### PR TITLE
Experimental pull from dataset-storage

### PR DESCRIPTION
**Pull request by rushi-r-s**

### CODE REVIEW

File: templates/datasets.html
Comment (Guptavivek27) [inline templates/datasets.html:1]: Need to add favorites and private section

**Review by Guptavivek27 — CHANGES_REQUESTED** [review]
Need to add favorites and private section

File: README.md [hunk 1]
```diff
@@ -68,12 +68,35 @@ python -m pip install -r requirements.txt
 
 ## How to Build and Run
 
-Migrate the Django database (only required to run once):
+Before starting the server, ensure that the Django database is correctly set up by performing migrations and creating a superuser for administrative access. Follow these steps:
+
+1. **Prepare Database Migrations**:  
+   Initialize database migrations needed for the Django models:
+```
+python manage.py makemigrations
```
Comment (alexandru-orhean) [inline README.md]: The makemigrations command is supposed to be called by the developers and not necessarily by the admins. Let's remove it for now.

File: README.md [hunk 2]
```diff
@@ -68,12 +68,35 @@ python -m pip install -r requirements.txt
...
+Create an administrative user to access the Django admin panel:
+```
+python manage.py createsuperuser
+```
+
+When prompted:
+- Username: `any_username` (choose any username you prefer)
+- Email Address: (can be left blank)
+- Password: `add_your_password` (enter a password of your choice)
+- Confirm the password by re-entering it. If prompted, you can press 'y' to bypass password validation, or re-enter the password if you prefer not to bypass.
```
Comment (alexandru-orhean) [inline README.md]: The createsuperuser command is self explanatory. No need to add extra information.

File: neurobazaar/services/datastorage/datastore_manager.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    
+    return wrapper
+
+class DatastoreManager:
+    def __init__(self):
+        self._datastores = {}
+    
+    @synchronized
+    def refresh(self):
+        print("Refreshing datastores...")
```
Comment (alexandru-orhean) [inline neurobazaar/services/datastorage/datastore_manager.py]: Replace the print statements with logging statements.

File: neurobazaar/services/datastorage/datastore_manager.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        self.refreshLocalFSDatastores()
+        self.refreshMongoDBDatastores()
+        print(f"Datastores after refresh: {self._datastores}")
+    
+    def refreshLocalFSDatastores(self):
+        records = {record.UUID: record for record in LocalFSDatastores.objects.all()}
+        print(f"LocalFSDatastores records: {records}")
```
Comment (alexandru-orhean) [inline neurobazaar/services/datastorage/datastore_manager.py]: This should be just a debugging message and is not supposed to print all the time.

File: neurobazaar/services/datastorage/datastore_manager.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        
+        # Create objects that exist in the model but not in the runtime
+        for datastoreUUID, record in records.items():
+            if datastoreUUID not in self._datastores:
+                self.addLocalFSDatastore(datastoreUUID, record.Directory_Path)
+
+        # # Remove objects that exist in the runtime but not in the model
+        # for datastoreUUID in list(self._datastores.keys()):
+        #     if datastoreUUID not in records:
+        #         del self._datastores[datastoreUUID]
```
Comment (alexandru-orhean) [inline neurobazaar/services/datastorage/datastore_manager.py]: Why is this code commented?

File: home/apps.py [hunk 6]
```diff
@@ -1,6 +1,35 @@
...
+                                                     Directory_Path=DATASTORE['default']['PATH'])
+
+                # Set all of the datastores to disconnected
+                from home.models import Datastores
+                Datastores.objects.update(Connected=False)
+                
+                # Create the datastore manager and initialize the datastores from the database
+                from neurobazaar.services.datastorage.datastore_manager import getDataStoreManager
+                manager = getDataStoreManager()
+                manager.refresh()
```
Comment (alexandru-orhean) [inline home/apps.py:30]: All of the datastores have to marked as connected after the refresh.

File: home/views.py [hunk 7]
```diff
@@ -23,265 +23,266 @@ def index(request):
...
-def register(request):
+def datastore(request):
+    username = request.user.username
+    # Handle form submissions for adding or removing datastores
     if request.method == 'POST':
-        user = UserCreationForm(request.POST)
-        if user.is_valid():
-            username = request.POST.get('username')
-            password = request.POST.get('password1')
+        if 'add_datastore' in request.POST:
```
Comment (alexandru-orhean) [inline home/views.py:30]: Need to mark datastore as connected after it get's added in the system

File: templates/datastore.html [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+                    <th>Remove Datastore</th>
+                </tr>
+            </thead>
+            <tbody>
+                {% for datastore in datastores %}
+                <tr>
+                    <td>{{ datastore.UUID }}</td>
+                    <td>{{ datastore.Name }}</td>
+                    <td>{{ datastore.Type }}</td>
+                    <td>{{ datastore.Created }}</td>
```
Comment (alexandru-orhean) [inline templates/datastore.html:100]: Need to add the connected column.

File: templates/datasets.html [hunk 9]
```diff
@@ -26,18 +26,28 @@ <h4 class="mb-3">Upload Datasets</h4>
...
+                        <select class="btn btn-outline-primary dropdown-toggle" name="repo" id="repo" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                            <option class="dropdown-item" value="public" selected>Public</option>
+                            <option class="dropdown-item" value="private">Private</option>
                         </select>
                     </div>
-                    <button type="submit" class="btn btn-primary">Upload</button>
+                    <div class="mb-3">
+                        <label class="m-2" for="datastore">Datastore:</label>
+                        <select class="btn btn-outline-primary dropdown-toggle" name="datastore" id="datastore" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                            {% for datastore in datastores %}
```
Comment (alexandru-orhean) [inline templates/datasets.html:45]: Only datastores that are connected can be used here

**Review by alexandru-orhean — CHANGES_REQUESTED** [review]
Revise and reply to all of the comments. Push all of the modifications and start another review.

File: neurobazaar/services/datastorage/datastore_manager.py (hunk 5 above)
Comment (rushi-r-s) [reply neurobazaar/services/datastorage/datastore_manager.py]: It kept deleting the other datastores that were required, but it was taken care as now it refreshes from the database and now we don't actually require this part of code.

File: templates/datasets.html
Comment (rushi-r-s) [reply templates/datasets.html:1]: Added from your repository

File: templates/datastore.html (hunk 8 above)
Comment (rushi-r-s) [reply templates/datastore.html:100]: Added the functionality to connect/disconnect form here.

File: home/views.py (hunk 7 above)
Comment (rushi-r-s) [reply home/views.py:30]: Taken care on line 46

File: home/views.py [hunk 10]
```diff
@@ -23,265 +19,258 @@ def index(request):
...
+            metadata = Datasets(
+                Username=user_instance,
+                Name=dname.name,
+                UUID=datasetid,
+                Datastore_UUID = datastore_instance,
+                Description=description,
+                Repository=repo,
+                Created=timezone.now().date(), 
+            )
+            metadata.save()
```
Comment (Guptavivek27) [inline home/views.py]: There is no metadata table.

File: templates/datasets.html [hunk 11]
```diff
@@ -142,76 +148,55 @@ <h4 class="mb-3">Upload Datasets</h4>
...
-                </a>
-                <div class="portfolio-caption">
-                    <div class="portfolio-caption-heading">{{ file }}</div>
-                    <div class="portfolio-caption-subheading text-muted">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer nec odio. Praesent libero. Sed cursus ante dapibus diam. Sed nisi.</div>
-                    <!-- Buttons -->
-                    <div class="d-flex align-items-center mt-2">
+        <div class="text-center">
+            <h2 class="section-heading text-uppercase">Datasets</h2>
+        </div>
+        <table class="table table-striped">
```
Comment (Guptavivek27) [inline templates/datasets.html]: Only public subgroup is formatted as table and not for private/favourites subgroup.

**Review by Guptavivek27 — CHANGES_REQUESTED** [review]

File: home/views.py (hunk 10 above)
Comment (rushi-r-s) [reply home/views.py]: metadata is the name of the variable here, the table name is correctly used as Datasets.

File: templates/datasets.html (hunk 11 above)
Comment (rushi-r-s) [reply templates/datasets.html]: Yes, me and professor worked on changing the models with correct name and adding the private key and foreign key. I pulled up from your existing branch. It is related to frontend and I believe you have to take care of it.

Comment (rushi-r-s) [reply templates/datasets.html]: Anyways, I myself made the changes in the html and in wrote appropriate code in views.py too. Now table is added. This issue is resolved

**Review by Guptavivek27 — COMMENTED** [review]
After my updates. I feel this branch is now to good to go.

**Review by Guptavivek27 — APPROVED** [review]

**Review by HuyNguyenDePaul — APPROVED** [review]
Approved, I have NOT verified or reviewed the code, but Rushi needs my approval to merge, I guess.

**Review by alexandru-orhean — APPROVED** [review]
Looks good to me!

**Pull request merged**